samza-core/src/main/java/org/apache/samza/storage/ChangelogStreamManager.java

```diff
@@ -0,0 +1,161 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.samza.SamzaException;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JavaStorageConfig;
+import org.apache.samza.config.JavaSystemConfig;
+import org.apache.samza.container.TaskName;
+import org.apache.samza.coordinator.stream.messages.CoordinatorStreamMessage;
+import org.apache.samza.coordinator.stream.messages.SetChangelogMapping;
+import org.apache.samza.coordinator.stream.CoordinatorStreamManager;
+import org.apache.samza.system.StreamSpec;
+import org.apache.samza.system.SystemAdmin;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.util.Util;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The Changelog manager is used to persist and read the changelog information from the coordinator stream.
+ */
+public class ChangelogStreamManager {
+
+  private static final Logger log = LoggerFactory.getLogger(ChangelogStreamManager.class);
+  private final CoordinatorStreamManager coordinatorStreamManager;
+
+  // This is legacy for changelog. Need to investigate what happens if you use a different source name
+  private static final String SOURCE = "JobModelManager";
+
+  /**
+   * Construct Changelog manager without a coordinator stream.
+   *
+   *  @param source Name of caller.
+   */
+  public ChangelogStreamManager(String source) {
+    coordinatorStreamManager = null;
+  }
+
+  /**
+   * Construct changelog manager with a bootstrapped coordinator stream.
+   *
+   * @param coordinatorStreamManager Coordinator stream manager.
+   */
+  public ChangelogStreamManager(CoordinatorStreamManager coordinatorStreamManager) {
+    this.coordinatorStreamManager = coordinatorStreamManager;
+  }
+
+  /**
+   * Read the taskName to partition mapping that is being maintained by this ChangelogManager
+   * @return TaskName to change log partition mapping, or an empty map if there were no messages.
+   */
+  public Map<TaskName, Integer> readPartitionMapping() {
+    log.debug("Reading changelog partition information");
+    final HashMap<TaskName, Integer> changelogMapping = new HashMap<>();
+    for (CoordinatorStreamMessage coordinatorStreamMessage : coordinatorStreamManager.getBootstrappedStream(SetChangelogMapping.TYPE)) {
+      SetChangelogMapping changelogMapEntry = new SetChangelogMapping(coordinatorStreamMessage);
+      changelogMapping.put(new TaskName(changelogMapEntry.getTaskName()), changelogMapEntry.getPartition());
+      log.debug("TaskName: {} is mapped to {}", changelogMapEntry.getTaskName(), changelogMapEntry.getPartition());
+    }
+    return changelogMapping;
+  }
+
+  /**
+   * Write the taskName to partition mapping.
+   * @param changelogEntries The entries that needs to be written to the coordinator stream, the map takes the taskName
+   *                       and it's corresponding changelog partition.
```

Comment:
Nitpick: formatting (align with "The" on the previousline.

Reply:
👍